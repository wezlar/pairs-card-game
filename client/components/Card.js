import React, { Component } from 'react';
import styled from 'styled-components';

const CardWrapper = styled.div`
  position: relative;
  width: 150;
  height: 217px;
  margin-right: 10px;
  margin-bottom: 10px;
`;

const CardBody = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  box-shadow: 3px 3px 3px rgba(0,0,0,0.2);
  transform-style: preserve-3d;
  transition: transform .3s ease-in-out;
`;

const CardBack = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 2;
  backface-visibility: hidden;
  background: #333;
`;

const CardFront = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  transition-delay: .15s;
  z-index: 1;
`;

class Card extends Component {
  render () {
    const { card: { image, name } } = this.props;

    const file = require(`../${image}`);

    return (
      <CardWrapper>
        <CardBody>
          <CardBack className="card__back">
            BACK { name }
          </CardBack>
          <CardFront className="card__face">
            <div dangerouslySetInnerHTML={ { __html: file } } />
          </CardFront>
        </CardBody>
      </CardWrapper>
    );
  }
};

export default Card;