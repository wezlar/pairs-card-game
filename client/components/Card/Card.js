import React, { memo } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const cardWidth = '138px';
const cardHeight = '200px';

const CardWrapper = styled.div`
  position: relative;
  width: ${cardWidth};
  height: ${cardHeight};
  margin-right: 10px;
  margin-bottom: 10px;
  perspective: 800px;
`;

const CardBody = styled.div`
  position: absolute;
  width: ${cardWidth};
  height: ${cardHeight};
  box-shadow: 3px 3px 3px rgba(0,0,0,0.2);
  transform-style: preserve-3d;
  transition: transform .3s ease-in-out;
  border-radius: 5px;
  transform: rotateY(0);
`;

const CardBack = styled.div`
  position: absolute;
  width: ${cardWidth};
  height: ${cardHeight};
  z-index: 2;
  backface-visibility: hidden;
  background-color:#fff;
  background-image: radial-gradient(closest-side, transparent 98%, red 99%),
    radial-gradient(closest-side, transparent 98%, black 99%);
  background-size: 30px 30px;
  background-position: 0 0, 15px 15px;
  border:solid 10px #fff;
  border-radius: 5px;
`;

const CardFront = styled.div`
  position: absolute;
  width: ${cardWidth};
  height: ${cardHeight};
  transition-delay: .15s;
  z-index: 0;
  border-radius: 5px;
`;

const SVGWrapper = styled.div`
  svg {
    width: ${cardWidth};
    height: ${cardHeight};
  }
`;

export function Card (props = {}) {
  const { 
    card: {
      image, 
      isFlipped,
      hasMatched,
    },
    onClick,
  } = props;

  const file = image 
    ? require(`!!raw-loader!../../${image}`)
    : '';

  const flippedStyle = isFlipped 
    ? { transform: 'rotateY(-180deg)' } 
    : {};
  
  const flippedFrontStyle = isFlipped 
      ? { transform: 'rotateY(180deg)' } 
      : {};
      
  const hasMatchedStyle = hasMatched 
    ? { boxShadow: '0 0 1px 2px #099f09' } 
    : { boxShadow: '0 0 1px 2px #00000000' };

  const cardFrontStyle = Object.assign({}, flippedFrontStyle, hasMatchedStyle);

  const cardFront = () =>
    <CardFront className="card__front" style={cardFrontStyle}>
      <SVGWrapper dangerouslySetInnerHTML={ { __html: file } } />
    </CardFront>;
  
  const cardBack = () => 
    <CardBack className="card__back" />;

  const cardBody = () => 
    <CardBody className="card__body" style={flippedStyle}>
      {cardBack()}
      {cardFront()}
    </CardBody>;

  // TODO: add hover state that slightly moves the card in opposite direction
  return <CardWrapper className="card__wrapper" onClick={onClick}>
      {cardBody()}
    </CardWrapper>;
};

Card.propTypes = {
  card: PropTypes.object,
  onClick: PropTypes.func,
};

Card.defaultProps = {
  card: {},
};

export default memo(Card);