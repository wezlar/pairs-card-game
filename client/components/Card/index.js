import React, { memo } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const cardWidth = 138;
const cardHeight = 200;

const CardWrapper = styled.div`
  position: relative;
  width: ${cardWidth + 10}px;
  height: ${cardHeight + 10}px;
  padding-top: 5px;
  padding-right: 5px;
  padding-bottom: 5px;
  padding-left: 5px;
  perspective: 800px;
  overflow: hidden;
`;

const CardBody = styled.div`
  position: absolute;
  width: ${cardWidth}px;
  height: ${cardHeight}px;
  box-shadow: 3px 3px 3px rgba(0,0,0,0.2);
  transform-style: preserve-3d;
  transition: transform .3s ease-in-out;
  border-radius: 5px;
  transform: rotateY(0);
`;

const CardBack = styled.div`
  position: absolute;
  width: ${cardWidth}px;
  height: ${cardHeight}px;
  z-index: 2;
  backface-visibility: hidden;
  background-color:#fff;
  background-image: radial-gradient(closest-side, transparent 98%, red 99%),
    radial-gradient(closest-side, transparent 98%, black 99%);
  background-size: 30px 30px;
  background-position: 0 0, 15px 15px;
  border:solid 10px #fff;
  border-radius: 5px;
`;

const CardFront = styled.div`
  position: absolute;
  width: ${cardWidth}px;
  height: ${cardHeight}px;
  transition-delay: .15s;
  z-index: 0;
  border-radius: 5px;
`;

const SVGWrapper = styled.div`
  svg {
    width: ${cardWidth}px;
    height: ${cardHeight}px;
  }
`;

export function Card (props = {}) {
  const { 
    card: {
      image, 
      isFlipped,
      hasMatched,
    },
    onClick,
  } = props;

  const file = image ? require(`!!raw-loader!../../${image}`) : '';
  const flippedStyle = isFlipped ? { transform: 'rotateY(-180deg)' } : {};
  const flippedFrontStyle = isFlipped ? { transform: 'rotateY(180deg)' } : {};
      
  const hasMatchedStyle = hasMatched 
    ? { boxShadow: '0 0 1px 2px #099f09' } 
    : { boxShadow: '0 0 1px 2px #00000000' };

  const cardFrontStyle = Object.assign({}, flippedFrontStyle, hasMatchedStyle);

  const cardFront = () =>
    <CardFront className="card__front" style={cardFrontStyle}>
      <SVGWrapper dangerouslySetInnerHTML={ { __html: file } } />
    </CardFront>;
  
  const cardBack = () => 
    <CardBack className="card__back" />;

  const cardBody = () => 
    <CardBody className="card__body" style={flippedStyle}>
      {cardBack()}
      {cardFront()}
    </CardBody>;

  // TODO: add hover state that slightly moves the card in opposite direction
  return <CardWrapper className="card__wrapper" onClick={onClick}>
      {cardBody()}
    </CardWrapper>;
};

Card.propTypes = {
  card: PropTypes.object,
  onClick: PropTypes.func,
};

Card.defaultProps = {
  card: {},
};

export default memo(Card);