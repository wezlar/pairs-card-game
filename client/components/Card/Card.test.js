import React from 'react'
import { shallow } from 'enzyme'
import { Card as Component } from './Card';

const mockCard = {
  cardIndex: 0,
  hasMatched: false,
  image: 'images/svg-cards/3_of_hearts.svg',
  isFlipped: false,
  name: '3 of hearts',
}

describe(`<Card />`, () => {
  test(`it doesn't blow up` , () => {
    const tree = shallow(<Component />)
    expect(tree.length).toEqual(1)
  });

  const mockOnClick = jest.fn();
  const props = {
    card: mockCard,
    onClick: mockOnClick,
  };

  test(`renders correctly unflipped`, () => {
    const card = shallow(<Component { ...props } />);
    expect(card.debug()).toMatchSnapshot();
  })

  test(`click is triggered`, () => {
    const card = shallow(<Component { ...props }  />);
    expect(mockOnClick).not.toHaveBeenCalled();
    card.find('.card__wrapper').simulate('click');
    expect(mockOnClick).toHaveBeenCalled();
  })
});