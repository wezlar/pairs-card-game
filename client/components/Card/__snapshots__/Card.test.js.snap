// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> renders correctly unflipped 1`] = `
"<styled.div className=\\"card__wrapper\\" onClick={[Function: mockConstructor]}>
  <styled.div className=\\"card__body\\" style={{...}}>
    <styled.div className=\\"card__back\\" />
    <styled.div className=\\"card__front\\" style={{...}}>
      <styled.div dangerouslySetInnerHTML={{...}} />
    </styled.div>
  </styled.div>
</styled.div>"
`;
